### DISCUSSION TITLE
How to configure kafka connect metrics ?

### DISCUSSION THREAD

**Discussion by raghulp18** [strimzi/strimzi-kafka-operator#10198] [General]
Hi Team,

I want to know how to configure a kafka connect metrics using a jmx and export it to prometheus ?
I have already gone through a configuration file of kafka-connect-metrics over here - https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/kafka-connect-metrics.yaml, but after adding this in the kafka-connect.yaml where can i check the metrics and how to export the port ? 

Below is the configuration file of the kafka-connect in my local.

apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaConnect
metadata:
  name: strimzi-kafka-connect-cluster
  annotations:
    strimzi.io/use-connector-resources: "true"
  namespace: kafka
spec:
  version: 3.6.1
  replicas: 1
  image: snowflake-jdbc-image
  bootstrapServers: my-cluster-kafka-bootstrap:9092
  metricsConfig:
    type: jmxPrometheusExporter
    valueFrom:
      configMapKeyRef:
        name: connect-metrics
        key: metrics-config.yml
  config:
    group.id: my-cluster
    offset.storage.topic: strimzi-connect-cluster-offsets
    config.storage.topic: strimzi-connect-cluster-configs
    status.storage.topic: strimzi-connect-cluster-status
    config.storage.replication.factor: 3
    offset.storage.replication.factor: 3
    status.storage.replication.factor: 3
    config.providers: directory
    config.providers.file.class: org.apache.kafka.common.config.provider.DirectoryConfigProvider

---
kind: ConfigMap
apiVersion: v1
metadata:
  name: kafka-metrics
  labels:
    app: strimzi
data:
  metrics-config.yml: |
    lowercaseOutputName: true
    lowercaseOutputLabelNames: true
    rules:
    #kafka.connect:type=connector-metrics,connector="{connector}"
    - pattern: 'kafka.connect<type=connector-metrics, connector=(.+)><>(connector-class|connector-type|connector-version|status): (.+)'
      name: kafka_connect_connector_$2
      value: 1
      labels:
        connector: "$1"
        $2: "$3"
      help: "Kafka Connect $2 JMX metric type connector"
      type: GAUGE

Please help here 

Thanks 
Raghul P

**Comment by scholzj** [comment]
You will have to format the YAMLs to make them readable. But in general, after you apply the YAML from the examples, the metrics will be exposed on every Connect Pod on port 9404 (path `/metrics/`).

**Reply by scholzj** [reply]
Then you probably didn't enabled it / used the example properly. But without you sharing the full YAMLs (including the `.status` section) and full operator logs, it is hard to comment in more detail about it.

**Reply by raghulp18** [reply]
## kafka.yaml

<img width="639" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/94948561/37cedce7-635c-42f4-8a24-d3c290b46b97">

## kafka-connect.yaml (metrics added over here)

<img width="657" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/94948561/43c8dedc-1d1b-409b-b7ba-7335bbc86602">

**Reply by raghulp18** [reply]
Can you please review it now ?

**Reply by scholzj** [reply]
As I said, you have to share the full resources as they are int he Kubernetes cluster - including their status sections. You also have to share the logs. Thanks.

**Reply by raghulp18** [reply]
[strimzi.zip](https://github.com/user-attachments/files/15711024/strimzi.zip)

@scholzj please find attached .zip file contains all yaml files and logs. 

And I'm able to connect to metrics on curl http://localhost:9404/metrics

<img width="1074" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/94948561/e056e4b7-ade3-4cb5-a2d5-7369de274e76">



Can you please help here on how to expose only the connector metrics to prometheus ?

**Reply by raghulp18** [reply]
Is there a way to change the metrics/prometheus metrics endpoint ?

**Reply by scholzj** [reply]
> And I'm able to connect to metrics on curl http://localhost:9404/metrics

Right, so that shows the metrics are working. To expose only some metrics, you would need to edit the rules in the ConfigMap. But I'm afraid I do not have any guide for how to do it and it might be also different for each connector (assuming you want some metrics produced by the connector itself).

**Reply by raghulp18** [reply]
Yeah, thanks for the help. But I want to know is there a way to change/overwrite the metrics to a different port ?

**Reply by raghulp18** [reply]
And where can I check all the available links under localhost:9404/metrics/ ?

**Reply by scholzj** [reply]
It is just the metrics, there is nothing else as far as I know.
